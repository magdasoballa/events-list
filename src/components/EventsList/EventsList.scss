.list-wrapper {
    display: flex;
    align-items: center;
    justify-content: center;
    h4 {
        display: flex;
        align-items: center;
        justify-content: center;
        font-family: "Lato", Arial, sans-serif;
    }
}
