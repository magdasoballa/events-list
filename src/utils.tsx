import { ReactElement } from "react";
import { act, render, RenderResult } from '@testing-library/react';

export const simpleRender = async (
    ui: ReactElement,
    { ...renderOptions } = {}
): Promise<RenderResult> => {
    let rendered: RenderResult = render(ui, { ...renderOptions });
    await act(async () => {
        rendered = render(ui, { ...renderOptions });
    });

    return rendered;
};
