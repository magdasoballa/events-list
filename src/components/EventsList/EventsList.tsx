import React, { useEffect, useState } from 'react';
import { List, ListItem, ListItemText, ListItemAvatar, Avatar, Typography, Pagination, Box, Select, MenuItem, FormControl, InputLabel, SelectChangeEvent, Button, CircularProgress } from '@mui/material';
import EventIcon from '@mui/icons-material/Event';
import { useNavigate } from 'react-router-dom';
import axios from 'axios';
import './EventsList.scss'
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../redux/store';
import { setEvents } from '../../redux/events/eventSlice';

const EventsList: React.FC = () => {
    const [currentPage, setCurrentPage] = useState<number>(1);
    const [itemsPerPage, setItemsPerPage] = useState<number>(10);
    const [loading, setLoading] = useState<boolean>(true);
    const navigate = useNavigate();
    const dispatch = useDispatch();
    const events = useSelector((state: RootState) => state.events.events);

    useEffect(() => {
        const fetchEvents = async () => {
            try {
                const response = await axios.get('http://localhost:3001/events');
                if (response.status === 200) {
                    dispatch(setEvents(response.data));
                } else {
                    console.error('Network response was not ok');
                }
            } catch (error) {
                console.error('Error fetching events:', error);
            } finally {
                setLoading(false);
            }
        };

        fetchEvents();
    }, [dispatch]);

    const handlePageChange = (event: React.ChangeEvent<unknown>, page: number) => {
        setCurrentPage(page);
    };

    const handleItemsPerPageChange = (event: SelectChangeEvent<number>) => {
        setItemsPerPage(Number(event.target.value));
        setCurrentPage(1);
    };

    const getPaginatedEvents = () => {
        const startIndex = (currentPage - 1) * itemsPerPage;
        const endIndex = startIndex + itemsPerPage;
        return events.slice(startIndex, endIndex);
    };

    const handleItemClick = (id: number) => {
        navigate(`/events/${id}`);
    };

    const handleAddEvent = () => {
        navigate('/add-event');
    };

    return (
        <div className="list-wrapper">
            <Box sx={{ padding: 2 }}>
                <Typography variant="h4" gutterBottom>
                    Events List
                </Typography>

                <FormControl variant="outlined" sx={{ marginBottom: 2, display: 'flex' }}>
                    <InputLabel id="items-per-page-label">Items per page</InputLabel>
                    <Select
                        labelId="items-per-page-label"
                        value={itemsPerPage}
                        onChange={handleItemsPerPageChange}
                        label="Items per page"
                    >
                        <MenuItem value={5}>5</MenuItem>
                        <MenuItem value={10}>10</MenuItem>
                        <MenuItem value={20}>20</MenuItem>
                        <MenuItem value={50}>50</MenuItem>
                    </Select>
                </FormControl>

                {loading ? (
                    <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '50vh' }}>
                        <CircularProgress />
                    </Box>
                ) : (
                    <>
                        <List>
                            {getPaginatedEvents().map(event => (
                                <ListItem key={event.id} alignItems="flex-start" button onClick={() => handleItemClick(event.id as number)}>
                                    <ListItemAvatar>
                                        <Avatar>
                                            <EventIcon />
                                        </Avatar>
                                    </ListItemAvatar>
                                    <ListItemText
                                        primary={event.title}
                                        secondary={
                                            <>
                                                <Typography
                                                    sx={{ display: 'inline' }}
                                                    component="span"
                                                    variant="body2"
                                                    color="text.primary"
                                                >
                                                    {event.date}
                                                </Typography>
                                                {" â€” " + event.description}
                                            </>
                                        }
                                    />
                                </ListItem>
                            ))}
                        </List>
                        <Button sx={{ width: '100%' }} variant="contained" color="primary" onClick={handleAddEvent}>
                            Add Event
                        </Button>
                        <Box sx={{ display: 'flex', justifyContent: 'center', marginTop: 2 }}>
                            <Pagination
                                count={Math.ceil(events.length / itemsPerPage)}
                                page={currentPage}
                                onChange={handlePageChange}
                                color="primary"
                            />
                        </Box>
                    </>
                )}
            </Box>
        </div>
    );
};

export default EventsList;
