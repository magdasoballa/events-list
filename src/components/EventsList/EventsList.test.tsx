import { screen } from '@testing-library/react';
import EventsList from './EventsList';
import axios from 'axios';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { simpleRender } from '../../utils';
import store from '../../redux/store';
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('EventsList', () => {
    afterEach(() => {
        jest.useRealTimers();
    });

    beforeEach(() => {
        mockedAxios.get.mockResolvedValue({
            status: 200,
            data: [
                { id: 1, title: 'Event 1', date: '2024-07-01', description: 'Description 1' },
                { id: 2, title: 'Event 2', date: '2024-07-02', description: 'Description 2' },
            ],
        });
    });

    test('should render event items', async () => {
        jest.useFakeTimers();
        const { container } = await simpleRender(
            <Provider store={store}>
                <Router>
                    <EventsList />
                </Router>
            </Provider>
        );

        expect(container).toMatchSnapshot();
        jest.advanceTimersByTime(30000);

        const lists = screen.getAllByText('Events List');
        expect(lists[0]).toBeInTheDocument();
    });
});
