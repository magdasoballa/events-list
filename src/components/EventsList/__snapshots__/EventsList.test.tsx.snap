// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsList should render event items 1`] = `
<div>
  <div
    class="list-wrapper"
  >
    <div
      class="MuiBox-root css-19midj6"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
      >
        Events List
      </h4>
      <div
        class="MuiFormControl-root css-co7lcf-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="items-per-page-label"
        >
          Items per page
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-controls=":r1:"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="items-per-page-label"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            role="combobox"
            tabindex="0"
          >
            10
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="10"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Items per page
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <ul
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      >
        <div
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button MuiListItem-alignItemsFlexStart css-1f57v5v-MuiButtonBase-root-MuiListItem-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-17o22dy-MuiAvatar-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EventIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1zm3 18H5V8h14z"
                />
              </svg>
            </div>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Event 1
            </span>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
              >
                2024-07-01
              </span>
               — Description 1
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button MuiListItem-alignItemsFlexStart css-1f57v5v-MuiButtonBase-root-MuiListItem-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart css-1m6607n-MuiListItemAvatar-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-17o22dy-MuiAvatar-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EventIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1zm3 18H5V8h14z"
                />
              </svg>
            </div>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Event 2
            </span>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 css-1phucj-MuiTypography-root"
              >
                2024-07-02
              </span>
               — Description 2
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </ul>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add Event
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiBox-root css-kl1xo3"
      >
        <nav
          aria-label="pagination navigation"
          class="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
        >
          <ul
            class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-colorPrimary MuiPaginationItem-textPrimary Mui-disabled MuiPaginationItem-previousNext css-1to7aaw-MuiButtonBase-root-MuiPaginationItem-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateBeforeIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-current="true"
                aria-label="page 1"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-colorPrimary MuiPaginationItem-textPrimary Mui-selected MuiPaginationItem-page css-1to7aaw-MuiButtonBase-root-MuiPaginationItem-root"
                tabindex="0"
                type="button"
              >
                1
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-colorPrimary MuiPaginationItem-textPrimary Mui-disabled MuiPaginationItem-previousNext css-1to7aaw-MuiButtonBase-root-MuiPaginationItem-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateNextIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;
